js/web/lib/wasm/jsep/webgpu/ops/argminmax.ts

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// TODO: this is the same naive implementation we use for reduce that has
+// performance limitations when the reduced axis is long. Need to add
+// a optimized codepath for this.
+
+import {DataType} from '../../../wasm-common';
+import {TensorView} from '../../tensor';
+import {ShapeUtil} from '../../util';
+import {AttributeWithCacheKey, createAttributeWithCacheKey} from '../attribute-with-cache-key';
+import {ComputeContext, GpuDataType, ProgramInfo, ProgramInfoLoader, ProgramMetadata} from '../types';
+
+import {createIndicesHelper, ShaderHelper} from './common';
+
+const validateInputs = (inputs: readonly TensorView[]): void => {
+  if (!inputs || inputs.length === 0 || inputs.length > 2) {
+    throw new Error('ArgMinMaxOp op requires 1 or 2 inputs.');
+  }
+  if (inputs[0].dataType !== DataType.float) {
+    throw new Error('Invalid input type.');
+  }
+};
+
+export interface ArgMinMaxAttributes extends AttributeWithCacheKey {
+  keepDims: boolean;
+  axes: number;
+  selectLastIndex: number;
+}
+
+type ArgMinMaxOp = (inputs: readonly TensorView[], axes: number[]) => string[];
+
+const createReduceProgramInfo =
+    (metadata: ProgramMetadata, inputs: readonly TensorView[], attributes: ArgMinMaxAttributes,
+     argMinMaxOp: ArgMinMaxOp): ProgramInfo => {
+      const outputShape: number[] = [];
+      const inputShape = inputs[0].dims;
+
```

Comment:
Can we try refactoring some code here?
We could generalize the same functions in reduce.ts and export so that we could use here instead of duplicating.

Reply:
yes, I think after merge we can try to unify and also add an optimized version of it.